bestiary/filters.py

```diff
@@ -100,29 +114,49 @@ def filter_skill_effects(self, queryset, name, value):
             # Filter effects based on effects of each individual skill. This ensures a monster will not show up unless it has
             # the desired effects on the same skill rather than across any skills.
             skills = Skill.objects.all()
+            skills_count = skills.count()
 
             for effect in value:
                 skills = skills.filter(skill_effect=effect)
 
             for pk in stat_scaling:
                 skills = skills.filter(scaling_stats=pk)
 
-            cooltime_filter = Q(cooltime__gte=min_cooltime) & Q(cooltime__lte=max_cooltime)
+            cooltime_filter = Q()
+
+            if max_cooltime is not None and max_cooltime > 0:
+                cooltime_filter &= Q(cooltime__lte=max_cooltime)
+            
+            if min_cooltime is not None and min_cooltime > 0:
+                cooltime_filter &= Q(cooltime__gte=min_cooltime)
+
             if min_cooltime == 0 or max_cooltime == 0:
-                cooltime_filter = Q(cooltime__isnull=True) | cooltime_filter
+                cooltime_filter |= Q(cooltime__isnull=True)
 
-            skills = skills.filter(
-                cooltime_filter,
-                hits__lte=max_num_hits,
-                hits__gte=min_num_hits,
-            )
+            if cooltime_filter:
+                skills = skills.filter(cooltime_filter)
+            
+            hits_filter = Q()
+
+            if max_num_hits:
+                hits_filter &= Q(hits__lte=max_num_hits)
+
+            if min_num_hits:
+                hits_filter &= Q(hits__gte=min_num_hits)
+
+            if hits_filter:
+                skills = skills.filter(hits_filter)
 
             if passive is not None:
                 skills = skills.filter(passive=passive)
 
             if aoe is not None:
                 skills = skills.filter(aoe=aoe)
 
+            # no skill filters
+            if skills_count == skills.count():
```

Comment:
Do you have any numbers for this specific optimization? Just want to make sure that adding 2 `count()` calls is worth the time to determine if we can skip the `filter(skills_in=skills)`

Reply:
Checked only for Bestiary, I expect similar results in Profile Monsters.

* Empty filter (`page=1`):
  * With double `count` check: 51ms whole db call
  * Without `count` check: 225ms whole db call (5 queries), `WHERE` clause has subquery: `WHERE ("bestiary_monster"."obtainable" = true AND "bestiary_monster_skills"."skill_id" IN (SELECT
"bestiary_skill"."id"
FROM "bestiary_skill"))`

* Full skill filter (no records returned, `buffs=7&buffs=13&buffs=70&debuffs=22&debuffs=23&debuffs=31&other_effects=51&other_effects=56&skills__scaling_stats__pk=24&skills__scaling_stats__pk=26&skills__passive=False&skills__aoe=True&skills__cooltime=3%2C8&skills__hits=2%2C6&leader_skill__attribute=2&leader_skill__attribute=5&leader_skill__area=2&leader_skill__area=5&page=1`):
  * With double `count` check: 226ms whole db call
  * Without `count` check: 200ms whole db call (203ms)

* Normal skill filter (Bomb debuff, `debuffs=23&page=1`):
  * With double `count` check: 50ms whole db call
  * Without `count` check: 33ms whole db call

Looks like it's quite a big deal for empty filter, but whenever we want to filter by anything from `skills`, then it's small downgrade. Of course I did only like 3 rounds per test, so it's not a big sample size.